### PR TITLE
Create distinct pre_analysis, analysis, and post_analysis phases

### PR DESCRIPTION
**Pull request by jpalmer37**
Fixes: https://github.com/BCCDC-PHL/auto-flu/issues/11

### CODE REVIEW

File: auto_flu/analysis.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    sequencing_run_id = run['sequencing_run_id']
+    
+    try:
+        logging.info(json.dumps({
+            "event_type": "analysis_started",
+            "sequencing_run_id": sequencing_run_id,
+            "pipeline_command": pipeline_command_str
+        }))
+        print(pipeline_command_str)
+        analysis_result = subprocess.run(pipeline_command_str, capture_output=True, check=True)
```
Comment (jpalmer37) [inline auto_flu/analysis.py:78]: This fails with a rather generic error: 

```
Traceback (most recent call last):
  File "/sharehome/john.palmer/git/auto-flu/auto_flu/analysis.py", line 79, in run_pipeline
    analysis_result = subprocess.run(pipeline_command_str, capture_output=True, check=True)
  File "/home/john.palmer/mamba/envs/test/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['nextflow', '-log', '/home/john.palmer/work/flu/results/test_analysis_results/240210_VH00502_167_AAF3VGCM5/fluviewer-nf-v0.3-output/240210_VH00502_167_AAF3VGCM5_fluviewer-nf_nextflow.log', 'run', 'BCCDC-PHL/fluviewer-nf', '-r', 'v0.3.0', '-profile', 'conda', '--cache', '/home/john.palmer/.conda/envs', '-work-dir', '/home/john.palmer/work/flu/work/work-240210_VH00502_167_AAF3VGCM5_fluviewer-nf_20240718095441', '-with-report', '/home/john.palmer/work/flu/results/test_analysis_results/240210_VH00502_167_AAF3VGCM5/fluviewer-nf-v0.3-output/240210_VH00502_167_AAF3VGCM5_fluviewer-nf_report.html', '-with-trace', '/home/john.palmer/work/flu/results/test_analysis_results/240210_VH00502_167_AAF3VGCM5/fluviewer-nf-v0.3-output/240210_VH00502_167_AAF3VGCM5_fluviewer-nf_trace.tsv', '-with-timeline', '/home/john.palmer/work/flu/results/test_analysis_results/240210_VH00502_167_AAF3VGCM5/fluviewer-nf-v0.3-output/240210_VH00502_167_AAF3VGCM5_fluviewer-nf_timeline.html', '--fastq_input', '/home/john.palmer/work/flu/fastq/test_fastq_symlinks/240210_VH00502_167_AAF3VGCM5', '--db', '/data/ref_databases/fluviewer/20230908/FluViewer_db.fa', '--outdir', '/home/john.palmer/work/flu/results/test_analysis_results/240210_VH00502_167_AAF3VGCM5/fluviewer-nf-v0.3-output']' returned non-zero exit status 1.
```
I copied that exact command to an interactive Python session, and it ran with no issues.

**Comment by jpalmer37** [comment]
@dfornika This isn't finalized yet; you'll notice there are still print / debugging statements. I've gotten stuck and could use a second opinion. See my comment above.

**Comment by jpalmer37** [comment]
Turns out the root cause was an active Conda environment interfering with the availability of Java that is used for Nextflow. Choosing a different Conda environment resolved the issue.

**Comment by jpalmer37** [comment]
Tested on a local mock setup and outputs matched the automated analysis production run.

**Review by dfornika — CHANGES_REQUESTED** [review]
Looking good! Just asking to re-name one function for consistency with other code-bases.

Edit: I re-considered, I think we should keep the function named the way you have it. We'll update the other code-bases to match this.

File: auto_flu/analysis.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    for flag, value in pipeline['pipeline_parameters'].items():
+        if value is None:
+            pipeline_command += ['--' + flag]
+        else:
+            pipeline_command += ['--' + flag, value]
+    
+    return pipeline_command
+
+
+def run_pipeline(config, pipeline, run):
```
[the remark this answers is not in the document]
Comment (dfornika) [reply auto_flu/analysis.py:49]: ...actually I take it back, I like your naming scheme. We can adopt this in the other tools to keep those consistent.

File: auto_flu/core.py [hunk 3]
```diff
@@ -62,49 +86,89 @@ def scan(config: dict[str, object]) -> Iterator[Optional[dict[str, object]]]:
...
+                "pipeline_name": pipeline['pipeline_name'],
+                "pipeline_version": pipeline['pipeline_version'],
+                "pipeline_dependencies": pipeline['dependencies'],
+                "sequencing_run_id": sequencing_run_id,
+                "conditions_checked": conditions_checked,
+            }))
+            continue
+
+        if pipeline:
+            analysis.run_pipeline(config, pipeline, run)
```
[the remark this answers is not in the document]
Comment (dfornika) [reply auto_flu/core.py:170]: ...nevermind, let's leave it as-is

**Review by dfornika — APPROVED** [review]
Looks good! If this is passing your test-runs then I'm good with merging this.

**Pull request merged**